contracts/src/Functions.sol

```diff
@@ -0,0 +1,178 @@
+
+library Functions {
+    using Address for address;
+    using SafeNativeTransfer for address payable;
+    using SafeTokenTransferFrom for IERC20;
+    using TokenInfoFunctions for TokenInfo;
+
+    error AgentDoesNotExist();
+    error InvalidToken();
+    error InvalidAmount();
+    error ChannelDoesNotExist();
+
+    function weth() internal view returns (address) {
+        AssetsStorage.Layout storage $ = AssetsStorage.layout();
+        return $.weth;
+    }
+
+    function ensureAgent(bytes32 agentID) internal view returns (address agent) {
+        agent = CoreStorage.layout().agents[agentID];
+        if (agent == address(0)) {
+            revert IGatewayBase.AgentDoesNotExist();
+        }
+    }
+
+    /// @dev Ensure that the specified parachain has a channel allocated
+    function ensureChannel(ChannelID channelID) internal view returns (Channel storage ch) {
+        ch = CoreStorage.layout().channels[channelID];
+        // A channel always has an agent specified.
+        if (ch.agent == address(0)) {
+            revert IGatewayV1.ChannelDoesNotExist();
+        }
+    }
+
+    /// @dev Invoke some code within an agent
+    function invokeOnAgent(address agent, address executor, bytes memory data)
+        internal
+        returns (bytes memory)
+    {
+        (bool success, bytes memory returndata) = (Agent(payable(agent)).invoke(executor, data));
+        return Call.verifyResult(success, returndata);
+    }
+
+    /// @dev transfer tokens from the sender to the specified agent
+    function transferToAgent(address agent, address token, address sender, uint128 amount)
+        internal
+    {
+        if (!token.isContract()) {
+            revert InvalidToken();
+        }
+
+        if (amount == 0) {
+            revert InvalidAmount();
+        }
+
+        IERC20(token).safeTransferFrom(sender, agent, amount);
+    }
+
+    function createAgent(bytes32 origin) internal {
+        CoreStorage.Layout storage core = CoreStorage.layout();
+        address agent = core.agents[origin];
+        if (agent == address(0)) {
+            agent = address(new Agent(origin));
+            core.agents[origin] = agent;
+            emit IGatewayBase.AgentCreated(origin, agent);
+        } else {
+            revert IGatewayBase.AgentAlreadyCreated();
+        }
+    }
+
+    /// @dev Transfer ether from an agent
+    function withdrawEther(
+        address executor,
+        address agent,
+        address payable recipient,
+        uint256 amount
+    ) internal {
+        bytes memory call = abi.encodeCall(AgentExecutor.transferNative, (recipient, amount));
+        invokeOnAgent(agent, executor, call);
+    }
+
+    function withdrawWrappedEther(
+        address executor,
+        address agent,
+        address payable recipient,
+        uint128 amount
+    ) internal {
+        withdrawNativeToken(executor, agent, weth(), address(this), amount);
+        WETH9(payable(weth())).withdraw(amount);
```

Comment:
This won't work as Gateway is not allowed to receive Ether.
https://github.com/Snowfork/snowbridge/blob/4e17456e3d774b3176966abdb80e1323d027d9d8/contracts/src/GatewayProxy.sol#L40-L44

A potential fix in https://github.com/Snowfork/snowbridge/pull/1321/commits/9081585234d8d8f9e6d0a6dede0a9918f2c0b030

Reply:
Good catch! yeah plz merge in